src/components/apps/launch/validate.js

```diff
@@ -0,0 +1,291 @@
+/**
+ * @author psarando
+ *
+ * App Launch Form validation functions.
+ */
+import { getMessage } from "@cyverse-de/ui-lib";
+
+import { validateDiskResourceName } from "../../data/utils";
+
+import constants from "./constants";
+
+/**
+ * @param {*} value - The app parameter value to check.
+ * @return {boolean} True if `value` is falsey, with the exception that Numbers can be exactly 0, and Arrays should have at least 1 item.
+ */
+const isEmptyParamValue = (value) =>
+    (!value && value !== 0) || (Array.isArray(value) && value.length < 1);
+
+const validateUnixGlob = (pattern) => {
+    if (pattern && (pattern.startsWith("/") || pattern.indexOf("../") >= 0)) {
+        return getMessage("validationUnixGlob");
+    }
+
+    return null;
+};
+
+const validateText = ({ value, validators }) => {
+    let errorMsg = null;
+
+    if (validators?.length > 0) {
+        validators.forEach((validator) => {
+            if (validator.params?.length > 0) {
+                let validatorMsg = null;
+
+                switch (validator.type) {
+                    case constants.VALIDATOR_TYPE.REGEX:
+                        const regexPattern = validator.params[0];
+                        try {
+                            const regex = new RegExp(regexPattern);
+                            if (!regex.test(value)) {
+                                validatorMsg = getMessage("validationRegex", {
+                                    values: { regexPattern },
+                                });
+                            }
+                        } catch (invalidRegex) {}
+                        break;
+
+                    case constants.VALIDATOR_TYPE.CHARACTER_LIMIT:
+                        const limit = validator.params[0];
+                        if (value.length > limit) {
+                            validatorMsg = getMessage("validationCharLimit", {
+                                values: { limit },
+                            });
+                        }
+                        break;
+
+                    default:
+                        break;
+                }
+
+                if (validatorMsg) {
+                    errorMsg = validatorMsg;
```

Comment:
Yes, `validators` is an array. I did intentionally go with only displaying 1 validation error at a time, and which one displays depends on the order the app integrator adds them to the param.

The Number Params story has an Integer param with 3 validators, even if they're not very realistic. So in this story, the validations are arranged so that `Must be greater than 1` will display for numbers less than 2, then `Must be between 3 and 7` will display for 2, 8, and 9, then `Must be less than 10` for numbers 10 or higher.

Reply:
Got it. Thanks for the clarification and example!